File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/http/ExecutionService.java
package ch.rmy.android.http_shortcuts.http;

import android.app.IntentService;
import android.content.Intent;

import java.util.HashMap;

import ch.rmy.android.http_shortcuts.realm.Controller;
import ch.rmy.android.http_shortcuts.realm.models.PendingExecution;
import ch.rmy.android.http_shortcuts.realm.models.ResolvedVariable;
import ch.rmy.android.http_shortcuts.utils.Connectivity;
import ch.rmy.android.http_shortcuts.utils.IntentUtil;
import io.realm.RealmResults;

public class ExecutionService extends IntentService {

    private static final int INITIAL_DELAY = 1500;

    public ExecutionService() {
        super(ExecutionService.class.getName());
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        final Controller controller = new Controller();

        while (Connectivity.INSTANCE.isNetworkConnected(this)) {
            RealmResults<PendingExecution> pendingExecutions = controller.getShortcutsPendingExecution();
            if (pendingExecutions.isEmpty()) {
                break;
            }

            final PendingExecution pendingExecution = pendingExecutions.first();
            long id = pendingExecution.getShortcutId();
            HashMap<String, String> variableValues = new HashMap<>();
            for (ResolvedVariable resolvedVariable : pendingExecution.getResolvedVariables()) {
                variableValues.put(resolvedVariable.getKey(), resolvedVariable.getValue());
            }

            controller.removePendingExecution(pendingExecution);

            try {
                Thread.sleep(INITIAL_DELAY);
                executeShortcut(id, variableValues);
            } catch (InterruptedException e) {
                break;
            }
        }

        controller.destroy();
    }

    private void executeShortcut(long id, HashMap<String, String> variableValues) {
        Intent shortcutIntent = IntentUtil.INSTANCE.createIntent(this, id, variableValues);
        startActivity(shortcutIntent);
    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/http/ExecutionService.kt
package ch.rmy.android.http_shortcuts.http

import android.app.job.JobParameters
import android.app.job.JobService
import android.content.Context
import android.os.Build
import android.support.annotation.RequiresApi
import ch.rmy.android.http_shortcuts.activities.ExecuteActivity
import ch.rmy.android.http_shortcuts.realm.Controller
import ch.rmy.android.http_shortcuts.utils.Destroyer

@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class ExecutionService : JobService() {

    val context: Context
        get() = this

    private val destroyer = Destroyer()

    val controller: Controller by lazy {
        destroyer.own(Controller())
    }

    override fun onStartJob(params: JobParameters): Boolean {
        val shortcutId = params.jobId.toLong()
        val pendingExecution = controller.getShortcutPendingExecution(shortcutId) ?: return false
        val tryNumber = pendingExecution.tryNumber + 1
        val variableValues = pendingExecution.resolvedVariables
                .associate { variable -> variable.key to variable.value }
        controller.removePendingExecution(shortcutId)
                .doOnTerminate {
                    executeShortcut(shortcutId, variableValues, tryNumber)
                    jobFinished(params, false)
                }
                .subscribe()
        return true
    }

    override fun onStopJob(params: JobParameters) = false

    private fun executeShortcut(id: Long, variableValues: Map<String, String>, tryNumber: Int) {
        val shortcutIntent = ExecuteActivity.IntentBuilder(context, id)
                .variableValues(variableValues)
                .tryNumber(tryNumber)
                .build()
        startActivity(shortcutIntent)
    }

    override fun onCreate() {
        super.onCreate()
        Controller.init(applicationContext)
    }

    override fun onDestroy() {
        destroyer.destroy()
    }

}
